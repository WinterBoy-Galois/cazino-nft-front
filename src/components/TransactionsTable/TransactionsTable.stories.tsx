import React, { useEffect, useState } from 'react';
import { boolean, object } from '@storybook/addon-knobs';

import TransactionsTable from '.';
import { action } from '@storybook/addon-actions';
import { TableColumn } from './lib/tableColumn';

export default {
  title: 'Components/TransactionsTable',
  component: TransactionsTable,
  decorators: [
    (storyFn: () => React.ReactElement) => (
      <div style={{ padding: '2rem', backgroundColor: '#113352' }}>{storyFn()}</div>
    ),
  ],
};

const data = {
  columns: [
    { selector: 'status', name: '', maxWidth: '180px' },
    { selector: 'time', name: 'Time', maxWidth: '220px' },
    { selector: 'hash', name: 'Hash', grow: 1, hideAtBreakpoint: 'md' },
    { selector: 'amount', name: 'Amount', maxWidth: '100px' },
  ],
  data: [
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600330565809,
      hash: 'b7e19012965f28872384703600f0242d0676136350b4b5e2eeffe03367008d3f',
      amount: 0.001,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332337322,
      hash: 'b5c29e9c14e3ce241eefa0bd0a654762733cc8eb1fa0eb8b8160ae935d89b7d9',
      amount: 0.0001,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332447372,
      hash: '41f34559b3e4e4f27d83710974ea1a946a4360d23c0bc2af9b9fa557fa781c71',
      amount: 0.00003,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332547389,
      hash: 'c39ff857c394edc428b6761b588bcce04a98fbaf59821db61b6d2a8550b01813',
      amount: 0.00003,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332592402,
      hash: 'cb5012a248960003e833fee23b9b63e0213508aec1a9212f85acae55368fff0c',
      amount: 0.000031,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332652455,
      hash: '7a7e0611bc2c40288d5acc82b8b6122b361d5ee0c29641a5d9d74e1fcfd8c2fe',
      amount: 0.000032,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332712467,
      hash: 'e2a88a58dea9988d5b9dbe23f987d039dbd1cceb8476842d8895f9d2c7e97ad2',
      amount: 0.0001,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332777484,
      hash: 'fcc14bd6338e5fd8af8a93b93e44447cde98133fa84dc62579f0adfd329e0d32',
      amount: 0.000091,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332862506,
      hash: 'd14efff16fdc8d5b999277714e41c77b5899f108c4565ca3e79d6a0dc3b852b9',
      amount: 0.00042,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600333047549,
      hash: '2600d16550fb9348e70a4d15ac7c0de73eee18bec4e5290b9ee31d798fee215b',
      amount: 0.00008,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600330565809,
      hash: 'b7e19012965f28872384703600f0242d0676136350b4b5e2eeffe03367008d3f',
      amount: 0.001,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332337322,
      hash: 'b5c29e9c14e3ce241eefa0bd0a654762733cc8eb1fa0eb8b8160ae935d89b7d9',
      amount: 0.0001,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332447372,
      hash: '41f34559b3e4e4f27d83710974ea1a946a4360d23c0bc2af9b9fa557fa781c71',
      amount: 0.00003,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332547389,
      hash: 'c39ff857c394edc428b6761b588bcce04a98fbaf59821db61b6d2a8550b01813',
      amount: 0.00003,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332712467,
      hash: 'e2a88a58dea9988d5b9dbe23f987d039dbd1cceb8476842d8895f9d2c7e97ad2',
      amount: 0.0001,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332592402,
      hash: 'cb5012a248960003e833fee23b9b63e0213508aec1a9212f85acae55368fff0c',
      amount: 0.000031,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332652455,
      hash: '7a7e0611bc2c40288d5acc82b8b6122b361d5ee0c29641a5d9d74e1fcfd8c2fe',
      amount: 0.000032,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332862506,
      hash: 'd14efff16fdc8d5b999277714e41c77b5899f108c4565ca3e79d6a0dc3b852b9',
      amount: 0.00042,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600332777484,
      hash: 'fcc14bd6338e5fd8af8a93b93e44447cde98133fa84dc62579f0adfd329e0d32',
      amount: 0.000091,
    },
    {
      status: 'DEPOSIT_CONFIRMED',
      time: 1600333047549,
      hash: '2600d16550fb9348e70a4d15ac7c0de73eee18bec4e5290b9ee31d798fee215b',
      amount: 0.00008,
    },
  ],
  paginationPerPage: 5,
};

export const Default = () => {
  const rowData = object('Data', data.data);
  const [rowsPerPage, setRowsPerPage] = useState(data.paginationPerPage);
  const [currentPage, setCurrentPage] = useState(1);
  const [rows, setRows] = useState(
    rowData.slice((currentPage - 1) * rowsPerPage, currentPage * rowsPerPage)
  );

  useEffect(() => {
    setRows(rowData.slice((currentPage - 1) * rowsPerPage, currentPage * rowsPerPage));
  }, [rowData, currentPage, rowsPerPage]);

  return (
    <TransactionsTable
      columns={object('Columns', data.columns) as TableColumn<any>[]}
      data={rows}
      paginationTotalRows={data.data.length}
      onChangePage={page => {
        action('onChangePage');
        setCurrentPage(page);
      }}
      paginationPerPage={rowsPerPage}
      onChangeRowsPerPage={r => setRowsPerPage(r)}
      paginationRowsPerPageOptions={[5, 10, 15, 20, 25]}
      progressPending={boolean('Loading', false)}
    />
  );
};
