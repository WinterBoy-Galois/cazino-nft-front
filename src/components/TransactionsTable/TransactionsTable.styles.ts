export const styles = {
  table: {
    style: {
      backgroundColor: 'transparent',
    },
  },
  header: {
    style: {
      backgroundColor: 'transparent',
    },
  },
  subHeader: {
    style: {
      backgroundColor: 'transparent',
    },
  },
  headRow: {
    style: {
      backgroundColor: '#113352',
      borderBottom: 'none',
      padding: '0 8px',
    },
  },
  headCells: {
    style: {
      textTransform: 'uppercase',
      color: '#6b7a8e',
      fontSize: '15px',
      padding: '16px 8px',
    },
    inactiveSortStyle: {
      '&:focus': {
        color: '#6b7a8e',
      },
      '&:hover': {
        color: '#6b7a8e',
      },
    },
  },
  rows: {
    style: {
      backgroundColor: '#112B46',
      fontSize: '15px',
      color: '#fff',
      '&:not(:last-of-type)': {
        borderBottom: 'none',
      },
      borderRadius: '8px',
      minHeight: '58px',
      '@media screen and (max-width: 768px)': {
        minHeight: '40px',
      },
      padding: '0 8px',
      '&:hover': {
        cursor: 'pointer',
      },
    },
    stripedStyle: {
      color: '#fff',
      backgroundColor: '#113352',
      borderRadius: '0',
    },
  },
  cells: {
    style: {
      paddingLeft: '8px',
      paddingRight: '8px',
    },
  },
  pagination: {
    style: {
      backgroundColor: 'transparent',
      color: '#5589FF',
      justifyContent: 'center',
      paddingTop: '16px',
      fontSize: '14px',
      border: 'none',
    },
    pageButtonsStyle: {
      color: '#5589FF',
      fill: '#5589FF',
      borderRadius: 0,
      '&:disabled': {
        color: 'rgba(85,137,255, 0.2)',
        fill: 'rgba(85,137,255, 0.2)',
      },
      '&:hover:not(:disabled)': {
        backgroundColor: 'transparent',
      },
      '&:focus': {
        backgroundColor: 'transparent',
      },
    },
  },
  noData: {
    style: {
      color: '#fff',
      backgroundColor: 'transparent',
      minHeight: '300px',
    },
  },
  progress: {
    style: {
      color: '#fff',
      backgroundColor: 'transparent',
      minHeight: '300px',
    },
  },
};
