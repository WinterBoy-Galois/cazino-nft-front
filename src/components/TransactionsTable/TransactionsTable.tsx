import React, { useMemo } from 'react';
import DataTable from 'react-data-table-component';
import { useTranslation } from 'react-i18next';
import { breakpoints } from '../../hooks/useBreakpoint.hook';
import Loading from '../Loading';
import { TableColumn } from './lib/tableColumn';
import { styles } from './TransactionsTable.styles';

interface IProps<T> {
  columns: TableColumn<T>[];
  data: T[];
  paginationTotalRows?: number;
  onChangePage?: (page: number, totalRows: number) => void;
  paginationPerPage?: number;
  onChangeRowsPerPage?: (rowsPerPage: number) => void;
  paginationRowsPerPageOptions?: number[];
  progressPending?: boolean;
  onRowClicked?: (row: T, event: MouseEvent) => void;
}

function TransactionsTable<T>(props: IProps<T>) {
  const { t } = useTranslation(['transactions']);

  const columns = useMemo(
    () =>
      props.columns.map(c => {
        const bp = breakpoints.find(b => b.name === c.hideAtBreakpoint);

        if (!bp) {
          return c;
        }

        return { ...c, hide: bp.minWidth };
      }),
    [props.columns]
  );

  return (
    <DataTable<T>
      {...props}
      pagination
      noHeader
      paginationServer
      customStyles={styles}
      striped
      columns={columns}
      paginationComponentOptions={{
        rowsPerPageText: t('table.rowsPerPage'),
        rangeSeparatorText: t('table.rangeSeparator'),
      }}
      noDataComponent={t('common:errors.NO_DATA')}
      progressComponent={
        <div className="w-100 h-100">
          <Loading />
        </div>
      }
    />
  );
}

export default TransactionsTable;
