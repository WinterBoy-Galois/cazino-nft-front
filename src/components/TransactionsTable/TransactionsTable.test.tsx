import React from 'react';
import { render } from '@testing-library/react';
import TransactionsTable from '.';

describe('TransactionsTable', () => {
  it.skip('should match snapshot', () => {
    // Arrange

    // Act
    const container = render(
      <TransactionsTable
        columns={[{ name: 'test' }]}
        data={[{ test: 'test' }]}
        paginationTotalRows={1000}
      />
    );

    // Assert
    expect(container).toMatchSnapshot();
  });
});
