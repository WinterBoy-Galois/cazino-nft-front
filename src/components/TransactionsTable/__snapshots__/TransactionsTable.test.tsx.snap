// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsTable should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-hHEjAm iECIgJ"
      >
        <div
          class="sc-kfYqjs ezVbOX"
        >
          <div
            class="sc-bdnylx ejLVfJ rdt_Table"
            role="table"
          >
            <div
              class="sc-gtssRu gfcMwa rdt_TableHead"
              role="rowgroup"
            >
              <div
                class="sc-dlnjPT jcwclA rdt_TableHeadRow"
                role="row"
              >
                <div
                  class="sc-hKFyIo sc-eCApGN sc-iqAbSa jPjSnc gcXkZv dwHiXA rdt_TableCol"
                >
                  <div
                    class="sc-crzoUp bwfVCA rdt_TableCol_Sortable"
                    id="column-undefined"
                    role="columnheader"
                    tabindex="0"
                  >
                    <div>
                      test
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sc-dIvqjp exodlN rdt_TableBody"
              offset="250px"
              role="rowgroup"
            >
              <div
                class="sc-jrsJCI HwxQx rdt_TableRow"
                id="row-0"
                role="row"
              >
                <div
                  class="sc-hKFyIo sc-eCApGN sc-jSFkmK dgFFWe gcXkZv cBZXUy rdt_TableCell"
                  data-tag="___react-data-table-allow-propagation___"
                  id="cell-QzpXd8z6r-undefined"
                  role="gridcell"
                >
                  <div
                    data-tag="___react-data-table-allow-propagation___"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <nav
          class="sc-lmgQde iKPurz rdt_Pagination"
        >
          <span
            class="sc-ezzayL sc-kLokih fGwysm gckGBs"
          >
            Rows per page:
          </span>
          <div
            class="sc-cxNIbT cGxHBE"
          >
            <select
              aria-label="Rows per page:"
              class="sc-iwaifL jzCzGz"
            >
              <option
                selected=""
                value="10"
              >
                10
              </option>
              <option
                value="15"
              >
                15
              </option>
              <option
                value="20"
              >
                20
              </option>
              <option
                value="25"
              >
                25
              </option>
              <option
                value="30"
              >
                30
              </option>
            </select>
            <svg
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
          <span
            class="sc-ezzayL sc-bYwzba fGwysm iGZUwt"
          >
            1-10 of 1000
          </span>
          <div
            class="sc-giAqnE bCPDDy"
          >
            <button
              aria-disabled="true"
              aria-label="First Page"
              class="sc-iJCRLp ljtHPi"
              disabled=""
              id="pagination-first-page"
              type="button"
            >
              <svg
                aria-hidden="true"
                height="24"
                role="presentation"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                />
                <path
                  d="M24 24H0V0h24v24z"
                  fill="none"
                />
              </svg>
            </button>
            <button
              aria-disabled="true"
              aria-label="Previous Page"
              class="sc-iJCRLp ljtHPi"
              disabled=""
              id="pagination-previous-page"
              type="button"
            >
              <svg
                aria-hidden="true"
                height="24"
                role="presentation"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </button>
            <button
              aria-disabled="false"
              aria-label="Next Page"
              class="sc-iJCRLp ljtHPi"
              id="pagination-next-page"
              type="button"
            >
              <svg
                aria-hidden="true"
                height="24"
                role="presentation"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </button>
            <button
              aria-disabled="false"
              aria-label="Last Page"
              class="sc-iJCRLp ljtHPi"
              id="pagination-last-page"
              type="button"
            >
              <svg
                aria-hidden="true"
                height="24"
                role="presentation"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                />
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
              </svg>
            </button>
          </div>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-hHEjAm iECIgJ"
    >
      <div
        class="sc-kfYqjs ezVbOX"
      >
        <div
          class="sc-bdnylx ejLVfJ rdt_Table"
          role="table"
        >
          <div
            class="sc-gtssRu gfcMwa rdt_TableHead"
            role="rowgroup"
          >
            <div
              class="sc-dlnjPT jcwclA rdt_TableHeadRow"
              role="row"
            >
              <div
                class="sc-hKFyIo sc-eCApGN sc-iqAbSa jPjSnc gcXkZv dwHiXA rdt_TableCol"
              >
                <div
                  class="sc-crzoUp bwfVCA rdt_TableCol_Sortable"
                  id="column-undefined"
                  role="columnheader"
                  tabindex="0"
                >
                  <div>
                    test
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-dIvqjp exodlN rdt_TableBody"
            offset="250px"
            role="rowgroup"
          >
            <div
              class="sc-jrsJCI HwxQx rdt_TableRow"
              id="row-0"
              role="row"
            >
              <div
                class="sc-hKFyIo sc-eCApGN sc-jSFkmK dgFFWe gcXkZv cBZXUy rdt_TableCell"
                data-tag="___react-data-table-allow-propagation___"
                id="cell-QzpXd8z6r-undefined"
                role="gridcell"
              >
                <div
                  data-tag="___react-data-table-allow-propagation___"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <nav
        class="sc-lmgQde iKPurz rdt_Pagination"
      >
        <span
          class="sc-ezzayL sc-kLokih fGwysm gckGBs"
        >
          Rows per page:
        </span>
        <div
          class="sc-cxNIbT cGxHBE"
        >
          <select
            aria-label="Rows per page:"
            class="sc-iwaifL jzCzGz"
          >
            <option
              selected=""
              value="10"
            >
              10
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="20"
            >
              20
            </option>
            <option
              value="25"
            >
              25
            </option>
            <option
              value="30"
            >
              30
            </option>
          </select>
          <svg
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
        <span
          class="sc-ezzayL sc-bYwzba fGwysm iGZUwt"
        >
          1-10 of 1000
        </span>
        <div
          class="sc-giAqnE bCPDDy"
        >
          <button
            aria-disabled="true"
            aria-label="First Page"
            class="sc-iJCRLp ljtHPi"
            disabled=""
            id="pagination-first-page"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
              />
              <path
                d="M24 24H0V0h24v24z"
                fill="none"
              />
            </svg>
          </button>
          <button
            aria-disabled="true"
            aria-label="Previous Page"
            class="sc-iJCRLp ljtHPi"
            disabled=""
            id="pagination-previous-page"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </button>
          <button
            aria-disabled="false"
            aria-label="Next Page"
            class="sc-iJCRLp ljtHPi"
            id="pagination-next-page"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </button>
          <button
            aria-disabled="false"
            aria-label="Last Page"
            class="sc-iJCRLp ljtHPi"
            id="pagination-last-page"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
              />
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
            </svg>
          </button>
        </div>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
